src/NuGet.Core/NuGet.ProjectModel/CacheFileFormat.cs

```diff
@@ -35,8 +34,9 @@ private static CacheFile Read(TextReader reader, ILogger log, string path)
         {
             try
             {
-                var json = JsonUtility.LoadJson(reader);
-                var cacheFile = ReadCacheFile(json);
+                string jsonString = reader.ReadToEnd();
+                var json = JsonDocument.Parse(jsonString);
```

Comment:
Unfortunately, NuGet was reading/writing json files in the worst possible way, and doing a "direct" port to System.Text.Json provides only limited benefit.  Especially, since System.Text.Json doesn't support `TextReader`, and this commit is working around that by loading the entire file into a single string, and this could actually make the perf worse, rather than better.

The problem with `JsonDocument` (similar to `JObject` from Newtonsoft.Json), is that it causes the entire json document to be loaded in memory at once, in the parser's own data structures. Since the it's first loading the entire file as a string, that's twice the file is loaded in memory, just two different data structures. And then converting it to our `CacheFile` data structure is the 3rd time it's loaded in memory.

It might not be a "huge" problem with a small file like `CacheFile`, but it would be a very significant problem for larger files.

It's much more efficient to use `JsonSerializer` instead to "stream" the json parsing, plus it's much less code on our side too.


Basically, we should follow the guidance that System.Text.Json provide: https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/deserialization

We can use `JsonSerializer.Deserialize<CacheFile>(fileStream)`  (delete the method that uses a `TextReader`), and to write use `JsonSerializer.Serialize(cacheFile, options)`, without calling `GetCacheFile`. Instead, use `[JsonPropertyName` attributes to map json and class property names when they don't match.  Then most of the code in this file can be deleted, in addition to making it have better runtime perf.

Reply:
I ported something else to System.Text.Json to demonstrate how it can be done:
* https://github.com/NuGet/NuGet.Client/pull/6083

As you can see, there's no need to write any code that know what the json property names are. Let the serializer handle that automatically. Also, no need for the temporary anonymous object on write.

Reply:
That was helpful! Thank you! I made sure my code followed suit as well. I used `sonSerializer.DeserializeAsync<CacheFile>(utf8Json: stream, SerializerOptions).GetAwaiter().GetResult();`, but I noticed you were able to use `Deserialize` overload by referencing the s.t.j package

Reply:
Wondering if it could lead into  a dead lock